### PR TITLE
Remove extra check that the system dll exists

### LINKED ISSUE #9884: Powershell .net class loads slow since MS-Security Update April 2019
Since May 2019 Security Rollup for Windows Server 2016, access to .net object via powershell is very slow, when a powershell variable breakpoint is set in the powershell.

For example loading speed of [System.Environment] takes over ten times with breakpoint set against no breakpoint set.

Since the behavior only occurs with latest security updates installed (May 2019 (KB4494440) or June 2019 (KB4503267)) i think it is related to the update.

Did some troubleshooting with procmon until now. For me it looks like when PSBreakpoint is set, there are multiple reads from `HKLM\System\CurrentControlSet\Control\Session Manager\Environment\__PSLockdownPolicy` can't repoduce this when there is no breakpoint set and can't reproduce this effect without one of the updates installed.

Following powershell code runs more than 10 times slower when i install Windows Security Rollup update Mai / June 2019 on my Windows Server 2016.
```

$measure1 = measure-command{
  foreach($i in 1..10000){
    [System.Environment]
  }
} |select -ExpandProperty TotalSeconds

Set-PSBreakpoint -Variable "test" -Action {$null}

$measure2 = measure-command{
  foreach($i in 1..10000){
    [System.Environment]
  }
} |select -ExpandProperty TotalSeconds

Write-Host "First measurement took $measure1 seconds, Second measurement took $measure2"

>>First measurement took 0.0482474 seconds, Second measurement took 0.8722697
```
Tested on:
                                                                            
PSVersion: 5.1.14393.2969                                                                          
PSEdition: Desktop                                                                                 
PSCompatibleVersions: {1.0, 2.0, 3.0, 4.0...}                                                                 
BuildVersion:10.0.14393.2969                                                                         
CLRVersion:4.0.30319.42000                                                                         
WSManStackVersion:3.0                                                                                     
PSRemotingProtocolVersion:2.3                                                                                     
SerializationVersion:1.1.0.1                                                                                 

Can anyone observe this behavior as well?

Do you have any suggestions to workaround this problem?

Is it related to some settings or installed software?

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Related #9884 

Remove extra check that the system dll exists. We need only one check, which is performed during first attempt to load this assembly.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/security/wldpNativeMethods.cs [hunk 1]
```diff
@@ -113,9 +113,8 @@ public static SystemEnforcementMode GetLockdownPolicy(string path, SafeHandle ha
         private static SystemEnforcementMode GetWldpPolicy(string path, SafeHandle handle)
         {
             // If the WLDP assembly is missing (such as windows 7 or down OS), return default/None to skip WLDP validation
-            if (s_hadMissingWldpAssembly || !IO.File.Exists(IO.Path.Combine(Environment.SystemDirectory, "wldp.dll")))
+            if (s_hadMissingWldpAssembly)
```
Comment (TravisEz13) [inline src/System.Management.Automation/security/wldpNativeMethods.cs]: This is still needed for Operating System before Windows 10

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 1 above)
Comment (TravisEz13) [reply src/System.Management.Automation/security/wldpNativeMethods.cs]: cc @PaulHigin

Comment (PaulHigin) [reply src/System.Management.Automation/security/wldpNativeMethods.cs]: Actually I feel this is the right fix.  We catch the *DllNotFoundException* and set the `s_hadMissingWldpAssembly` field there.  I don't see a need to check it every time at the top of the method.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Review by TravisEz13 — APPROVED** [review]

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/security/wldpNativeMethods.cs]: Yes, it is already in the PR description. 

I just do not understand why the security fix (mentioned in related issue) led to this problem if it is a long-standing code? Or is there still another problem?

**Comment by daxian-dbw** [comment]
@iSazonov Is P/Invoke in general becomes slower or only the P/Invoke calls to `wldp.dll` becomes slower?
Another question, the trace you collected, are they from Windows PowerShell 5.1 or the latest PowerShell Core?

**Comment by PaulHigin** [comment]
@iSazonov  I can contact the wldp team but I need more information.  Can you clearly summarize how you ran your perf tests and obtained the traces?  I don't know what 'with bp' and 'without bp' means...  We can't cache filepath information and need to rely on wldp APIs to be performant.  Thanks!

**Comment by daxian-dbw** [comment]
> I don't know what 'with bp' and 'without bp' means...

I believe it's `with breakpoint` and `without breakpoint`.

**Comment by PaulHigin** [comment]
I took a look at the original issue (#9884), and found the problem.   The issue is not related to the wldp.dll, however I still feel this fix is correct and should be taken.

The problem is the debugger disabling work that went into the current preview builds.  The script debugger is now disabled if the system is locked down, and any attempt to change debugger state is guarded by a system lock down check.  System lock down state is normally cached, but for the (normal) case where no policy is applied then a check is always performed for debug hooks (__PSLockdownPolicy) for testing purposes, and these checks include environment and registry settings.  Note that this perf problem only occurs when a debug breakpoint is set.

I'll work on a fix for this and create a new PR.

**Comment by iSazonov** [comment]
I used a test script from original issue:
File c:\temp\b.ps1
```powershell
Get-PSBreakpoint | Remove-PSBreakpoint
$sw = [System.Diagnostics.StopWatch]::startnew()
#"Without bp:`t$($sw.Elapsed.Ticks)"
$null = Set-PSBreakpoint -Line 2 -Script $PSCommandPath
$sw = [System.Diagnostics.StopWatch]::startnew()
#"With bp:`t$($sw.Elapsed.Ticks)"
```
You can uncomment lines to see elapsed times.

Test with breakpoint:
1. Run pwsh
2. Run PerfView
3. Start collect in PerfVew
4. Run in PowerShell Core session:
```powershell
for ($i = 0 ; $i -lt 1000; $i++) {
     c:\temp\b.ps1
}
```
5. Wait for the end of the script block and then stop collect

Now comment "$null = Set-PSBreakpoint -Line 2 -Script $PSCommandPath" line in the test file and repeat the test steps  for testing without breakpoint.

**Comment by iSazonov** [comment]
@daxian-dbw 
> Is P/Invoke in general becomes slower or only the P/Invoke calls to wldp.dll becomes slower?

I don't know. I guess if P/Invoke were the case then we would see a slowdown in other places too.

> the trace you collected, are they from Windows PowerShell 5.1 or the latest PowerShell Core?

It is for latest PowerShell Core (my local build - I collect for debug (to see code paths) and release (to see real delays)). 

> how much improvement do you get with this change?

Visually the test script (b.ps1 file) runs ~3x faster in with-breakpoint case but still 4x slower than without-breakpoint case.

**Comment by iSazonov** [comment]
@PaulHigin 
>I took a look at the original issue (#9884), and found the problem. The issue is not related to the wldp.dll, however I still feel this fix is correct and should be taken.
The problem is the debugger disabling work that went into the current preview builds.

It is not clear why mentioned Windows security fix cause the original issue.  It seems you found another issue. 
Trace for the PR build shows
image
 that we spend a lot of time in  SecuritySupport.GetSaferPolicy() that is advapi32.SaferIdentifyLevel P/Invoke. Perhaps the security patch did slow down the API, not wldp.dll.

**Comment by iSazonov** [comment]
I look `Get-Module -ListAvailable -All` - it seems it is slower now too.
Also we have #10264.

**Comment by PaulHigin** [comment]
@iSazonov  No, the issue I found is the same as #9884.  I simply repro'd that issue (setting debug break point causes serious perf problem) and found it was a regression from the debugger disabling change.  I think is also a PInvoke problem as well, apparently due to the security fix, but this particular issue is due to a regression caused by a previous code change.

I'll create a PR with the fix.

**Comment by iSazonov** [comment]
@TravisEz13 The PR is ready to merge.

**Pull request merged**